### PR TITLE
Implement Magic Wand Tool (#128)

### PR DESCRIPTION
**Pull request by HenryJia**
This implements the magic wand tool of #128

### CODE REVIEW

File: src/tiled/magicwandtool.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+{
+    MapDocument *document = mapDocument();
+
+    if (mSelectedRegion != document->selectedArea()) {
+        QUndoCommand *cmd = new ChangeSelectedArea(document, mSelectedRegion);
+        document->undoStack()->push(cmd);
+    }
+
+    brushItem()->setTileRegion(mSelectedRegion);
+    document->setSelectedArea(mSelectedRegion);
```
Comment (bjorn) [inline src/tiled/magicwandtool.cpp]: This line seems wrong. The selected area is already applied to the document by the `ChangeSelectedArea` undo command created and pushed on the undo stack above.

Also, this seems like the wrong moment to set the tile region of the brush item. Shouldn't that be done in the `tilePositionChanged` function to provide a preview of the area that's going to be selected?

File: src/tiled/magicwandtool.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+    if (!mStamp)
+        return;
+
+    for (int x = 0; x < mStamp->width(); x++)
+        for (int y = 0; y < mStamp->height(); y++)
+            if (!mStamp->cellAt(x, y).isEmpty())
+                mRandomList.append(mStamp->cellAt(x, y));
+}
+*/
```
Comment (bjorn) [inline src/tiled/magicwandtool.cpp]: Please remove all this commented out stuff. It's irrelevant to the magic wand tool.

**Comment by bjorn** [comment]
To fix the AppVeyor continue-integration build you should add `magicwandtool.cpp` and `magicwandtool.h` to the files list in `src/tiled/tiled.qbs`.

**Comment by HenryJia** [comment]
OK, Done.

File: src/tiled/magicwandtool.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    //void updateStatusInfo();
+
+private:
+    enum SelectionMode {
+        Replace,
+        Add,
+        Subtract,
+        Intersect
+    };
```
Comment (bjorn) [inline src/tiled/magicwandtool.h]: You have these selection modes here but you're not using them. If the tool is working so far, why not try and implement that functionality like in the `TileSelectionTool`?

**Comment by HenryJia** [comment]
OK

**Comment by HenryJia** [comment]
I'm not fully sure how the selection modes work in the actual program (as in how you change it)

**Comment by bjorn** [comment]
> I'm not fully sure how the selection modes work in the actual program (as in how you change it)

Please see if you can just derive this by reading the code in `TileSelectionTool`.

**Comment by HenryJia** [comment]
Oh I see, through modifiers on the keyboard.

**Comment by HenryJia** [comment]
I've added the keyboard modifiers. However the shift (add) modifiers doesn't work because computefillRegion only computes it within the selected region.

**Comment by bjorn** [comment]
> I've added the keyboard modifiers. However the shift (add) modifiers doesn't work because computefillRegion only computes it within the selected region.

Alright, that will require some changes to `computeFillRegion` then. Essentially, it needs to be optional whether it takes the selection into account because it does still need to do that for the fill tool. I will try to push a change for this sometime soon.

**Comment by HenryJia** [comment]
I forgot to say, the magic wand tool will also need a new icon for it. Currently I've just got the icon set to the same one as bucket fill.

**Comment by bjorn** [comment]
> I forgot to say, the magic wand tool will also need a new icon for it. Currently I've just got the icon set to the same one as bucket fill.

Fortunately, the GIMP also has a magic wand tool so we can just use the same icon. Most of the other icons are also taken from the GIMP.

**Comment by HenryJia** [comment]
Hmm, I don't know where to get the GIMP icons.

**Comment by toddcarnes** [comment]
You should be able to find the source at gimp.org

**Comment by HenryJia** [comment]
I've added the magic wand icon.

What shortcut key should I use or this?

File: src/tiled/magicwandtool.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    void mousePressed(QGraphicsSceneMouseEvent *event);
+    void mouseReleased(QGraphicsSceneMouseEvent *event);
+
+    void languageChanged();
+
+protected:
+    void tilePositionChanged(const QPoint &tilePos);
+
+    //void updateStatusInfo();
```
Comment (bjorn) [inline src/tiled/magicwandtool.h]: Please remove this commented line.

File: src/tiled/magicwandtool.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    MapDocument *document = mapDocument();
+
+    QRegion selection = document->selectedArea();
+
+    if(modifiers == Qt::ShiftModifier)
+        selection += mSelectedRegion;
+    else if(modifiers == Qt::ControlModifier)
+        selection -= mSelectedRegion;
+    else if(modifiers == (Qt::ControlModifier | Qt::ShiftModifier))
+        selection &= mSelectedRegion;
```
Comment (bjorn) [inline src/tiled/magicwandtool.cpp]: Please always put a space between the `if` and the `(`. It just happens to be my code style.

**Comment by HenryJia** [comment]
Done

**Comment by bjorn** [comment]
Thanks for your work on this tool! I've squashed your commits and pushed it as 57c27c1cf2107713ee9ead2c8504f485ffe7a3bb. I've then fixed the `computefillRegion` problem in change 4a6322f134e4d31923a46d259e5e80e8814d8231.

**Pull request closed without merging**